### PR TITLE
MAINT: don't warn for symbols needed by import_array()

### PR DESCRIPTION
**Pull request by ngoldbaum**
`_ARRAY_API` and `_UFUNC_API` are needed by the `import_array()` and `import_ufunc()` mechanisms, respectively. I don't think it's worthwhile to warn on accessing either from `np.core` since there's little users can do to fix this for old compiled code.

See https://github.com/astropy/astropy/pull/15495 where warning on these attributes using the old paths caused inscrutible CI breakage because erroring on this can leave behind a partially imported module or a partially imported dependency or transitive dependency.

ping @mtsokol since I pinged you on the astropy issue that this should fix.

### CODE REVIEW

**Review by pllim — APPROVED** [review]
Thanks!

**Review by mhvk — APPROVED** [review]
Confirmed that astropy tests pass with this PR, and it seems simple and clearly commented. A test would be nice but hard to do, so approving.

File: numpy/core/_multiarray_umath.py [hunk 1]
```diff
@@ -8,6 +8,12 @@
     if isinstance(attr, ufunc):
         globals()[item] = attr
 
+# The NumPy 1.x import_array() and import_ufunc() mechanisms expect
+# these symbols to be available without error. If an extension was compiled
+# against NumPy 1.x it will have these paths hard-coded.
+globals()['_ARRAY_API'] = _multiarray_umath._ARRAY_API
+globals()['_UFUNC_API'] = _multiarray_umath._UFUNC_API
```
Comment (seberg) [inline numpy/core/_multiarray_umath.py:15]:
```suggestion
_ARRAY_API = _multiarray_umath._ARRAY_API
_UFUNC_API = _multiarray_umath._UFUNC_API
```
No point in globals, no?

We need to break 1.x opening here soon enough, but I thought we kept the path in place also for 2.x.

**Comment by ngoldbaum** [comment]
Why does this pain need to be felt eventually? Can't we just keep these working forever, along with the other imports we need for pickles? This isn't so much about astropy compiling against older numpy, but any other downstream compiles code with the old path baked in that downstream can't really do anything about. What's the benefit for warning in those cases?

**Comment by seberg** [comment]
> Why does this pain need to be felt eventually? 

Because some things you may be doing when you compile against NumPy 1.x (and not also 2.x) will lead to crashes or undefined behavior.  This is already the case in principle, we just ignore it for simplicity.
If we do things like change the `PyArray_Descr` struct layout (which would be nice to bump the itemsize and to add more flag space for example), then such changes would be common enough that we can't just hope as we do currently.

So it's either forcing compiling with 2.x, not actually doing a major version bump in C at all, or just accepting limited amount of undefined behavior...

(Right now we are still in a situation that *if* you can also compile against 2.x, you are safe.  But we don't know that you can.)

**Comment by seberg** [comment]
And fair: changing the default integer doesn't require this strictly.  Just if you want to use our helper to know which one is the actual default integer, you will only have it if you compile against 2.x.

**Comment by mhvk** [comment]
@seberg - at the very least, the error message should be changed, i.e., these particular attributes should be special-cased in `__getattr__`. As is, the `np.core` deprecation is super obscure for compiled modules, since one's own code does not refer to `np.core.multiarray`, it just does `import_arrays()`. 

Though the astropy experience shows it is worse than that: the cause of the error was that one of our dependencies (`pyerfa`) is compiled against numpy 1.26, but actual traceback made it appear the error was in one of our internal ufuncs, even though in our numpy-dev run that is in fact compiled against numpy-dev. See https://github.com/astropy/astropy/pull/15495#issuecomment-1766834705. 

 I don't know what is the best way forward. A different warning message would certainly help, since then we can just filter on it until all numpy 2.0 is actually out and all dependencies compile against it, without also filtering warnings we should actually see and address to be compatible with numpy 2.0 ourselves. Better still would be if it were possible to implement a separate warning if something is compiled against the wrong numpy C api. (Indeed, even an error in that case may be fine.)

**Comment by pllim** [comment]
To ask all downstream packages to compile against unreleased numpy is unrealistic. As a downstream maintainer, I am hesitant to ship something that is compiled against what I know is unstable (other than the package I am touching itself). If anything breaks in the chain, it would be a pain to track every time numpy nightly wheel is updated. It is already painful enough on just the Python side in this regard the in "devdeps" CI job.

Also, even if I am adventurous and want to do that, it would make the build machinery super ugly. I will have to ask the build to look at Scientific Python wheels storage instead of the standard PyPI and so on. And if there are multiple things in the chain that are affected, I also have to beg them for a similar setup. And then, I have to remember *not* to do that at release time.

In astropy's case, only astropy itself and pyerfa are affected. But think about another package down-downstream that uses astropy, scipy, scikit-image, and a plethora of other packages with C-extensions calling NumPy. It would be a nightmare for them. And furthermore, usually those packages are maintained by someone not very well-versed with advanced packaging and their day job are to write papers, not packaging software.

So, I think for now, I will just ignore this warning in astropy and see if the tests would run...

Thank you for your time and consideration.

**Comment by seberg** [comment]
> (Indeed, even an error in that case may be fine.)

Right, there will be pain because at some point we need to make this an error.  We have two options:
1. Keep the warning, but I am very happy with an improved message.
2. Just make it an error now and live with the fact that it might be a week of bumpy nightlies while everyone makes sure to compile them with NumPy 2.x

I am absolutely :+1: to adding a custom error in `__getattr__()` whether it is an error right now, or whether we do it in a week.
The reason is that we will automatically get an error from inside C (but we currently lie in C).  But we have no control over the error message you will see.  OTOH, if we add the error in `__getattr__()` we can make sure it is a clear one (and can even keep updating it).

**Comment by seberg** [comment]
> To ask all downstream packages to compile against unreleased numpy is unrealistic. As a downstream maintainer, I am hesitant to ship something that is compiled against what I know is unstable (other than the package I am touching itself).

@pllim I am only asking about *nightlies* not actual releases?!  I don't really see a problem with compiling against 2.x for those?

One issue I do understand is that `pyerfa` might not have any nightlies right now, because it is so small, so I guess you might have to build it from the repository directly?!  (where it is OK to remove the `<2` assuming that you would re-add it if you cut a release before the NumPy 2.0 release.)

**Comment by ngoldbaum** [comment]
Pushed an update in case we decide to go with this instead of #24949.

**Comment by ngoldbaum** [comment]
> So it's either forcing compiling with 2.x, not actually doing a major version bump in C at all, or just accepting limited amount of undefined behavior...

Ah, I didn't realize that we're planning to break old C extensions. In that case this is unnecessary, but we definitely want to have better error messages in that case. That said, since we haven't yet made the changes that will break old C extensions in common cases this still might be needed if we decide that's not what we want to do.

**Comment by pllim** [comment]
> Downstream must ensure they build with 2.0 when they test against it or release nightlies that are supposed to work with it.

I cannot speak for all downstream, but seems reasonable for astropy though its maintainers need to debate it out first. Tom and Simon wanted the `numpy<2` pin and they are more knowledgeable than me in terms of building infrastructure. Thanks!

**Pull request merged**